Support for Backoff

euonymos opened the pull request:
## Summary

To handle automatically rate limit errors this PR allows clients to use exponential backoff with maximum delay threshold by optionally specifying these two options when creating an environment with `mkMaestroEnv`. This solves #39.

## Type of Change

Please mark the relevant option(s) for your pull request:

- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Documentation update (adding or updating documentation related to the project)

## Checklist

Please ensure that your pull request meets the following criteria:

- [X] I have read the [Contributing Guide](CONTRIBUTING.md)
- [X] My code follows the project's coding style and best practices
- [X] My code is appropriately commented and includes relevant documentation
- [X] I have added tests to cover my changes
- [X] All new and existing tests pass
- [X] I have updated the documentation, if necessary

## Testing

- [X] errors without backoff
- [X] works with default backoff settings

## TODO

- [ ] reformat code
- [ ] add MAESTRO_API_KEY in CI?

sourabhxyz commented:
Thanks a lot @euonymos for this contribution! I would have a thorough look at it tomorrow. At the same time, we have this PR: https://github.com/maestro-org/haskell-sdk/pull/33 which is what is being used for in some of the other tools, like [Atlas](https://github.com/geniusyield/atlas). If you would like, you can set target of this PR to `maestro-org:drop-v0`.

euonymos commented:
Hey @sourabhxyz! I added a pair of tests, though they require an API key, so they will fail in CI. Not sure what's the best way to handle this. I also rebased on the branch you pointed out, thank you.

sourabhxyz commented:
Thanks again @euonymos, I am merging this and would change few things slightly like changing the `_` prefix in say `_maeToken` as I moved away from template haskell towards using generics for lenses.
